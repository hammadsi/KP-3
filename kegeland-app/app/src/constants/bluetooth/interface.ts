import {DeviceStackParamList} from '~routes/interface';

export type BluetoothCharacteristic = string;
export type BluetoothService = {
  uuid: string;
  characteristics: BluetoothCharacteristic[];
};

export type BluetoothProfile = {
  navScreen: keyof DeviceStackParamList;
  services: string[];
  batteryService: {
    service: BluetoothService;
    fn: (bytes: number[]) => number;
  };
};

export type PeripheralNotification = {
  value: number[];
  peripheral: string;
  characteristic: string;
  service: string;
};
