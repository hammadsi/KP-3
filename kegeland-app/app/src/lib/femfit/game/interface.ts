import Matter from 'matter-js';
import {GameEngineUpdateEventOptionType} from 'react-native-game-engine';

export type Position = {
  x: number;
  y: number;
};

export type EntityBase = {
  body: Matter.Body;
};

export type ExerciseScheme = {
  name: string;
  description: string;
  repetitions: number;
  data: [number, number][];
};

export type GameEntities = {
  [key: string]: EntityBase & {[key: string]: any};
} & {
  physics: {
    engine: Matter.Engine;
    world: Matter.World;
  };
  exercise: ExerciseScheme & {currRep: number; currStep: number};
};

export type IGameEngineSystem = (
  entities: GameEntities,
  update: GameEngineUpdateEventOptionType,
) => GameEntities;
