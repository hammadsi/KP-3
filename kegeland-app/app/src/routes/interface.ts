import {
  CompositeScreenProps,
  NavigatorScreenParams,
} from '@react-navigation/native';
import type {BottomTabScreenProps} from '@react-navigation/bottom-tabs';
import type {StackScreenProps} from '@react-navigation/stack';

/**
 * Navigation interface for auth routes
 */
export type AuthStackParamList = {
  Login: undefined;
  Register: undefined;
  'Forgot password': undefined;
};

export type DeviceStackParamList = {
  Devices: undefined;
  Femfit: undefined;
};

/**
 * Navigation interface for settings routes
 */
export type SettingsStackParamList = {
  Settings: undefined;
  'Connect device': undefined;
};

/**
 * Base navigation interface
 */
export type RootTabParamList = {
  AuthStack: NavigatorScreenParams<AuthStackParamList> | undefined;
  DeviceStack: NavigatorScreenParams<DeviceStackParamList> | undefined;
  SettingsStack: NavigatorScreenParams<SettingsStackParamList> | undefined;
};

/**
 * Interface for screens on root-level
 */
export type RootScreenProps<
  RouteName extends keyof RootTabParamList = keyof RootTabParamList,
> = BottomTabScreenProps<RootTabParamList, RouteName>;

/**
 * Interface for screens under the auth-stack
 */
export type AuthScreenProps<
  RouteName extends keyof AuthStackParamList = keyof AuthStackParamList,
> = CompositeScreenProps<
  StackScreenProps<AuthStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

export type DeviceScreenProps<
  RouteName extends keyof DeviceStackParamList = keyof DeviceStackParamList,
> = CompositeScreenProps<
  StackScreenProps<DeviceStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

/**
 * Interface for screens under the settings-stack
 */
export type SettingsScreenProps<
  RouteName extends keyof SettingsStackParamList = keyof SettingsStackParamList,
> = CompositeScreenProps<
  StackScreenProps<SettingsStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;
