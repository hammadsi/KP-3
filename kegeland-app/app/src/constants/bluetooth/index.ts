import FEMFIT_PROFILE from '~lib/femfit/bluetooth/profile';

import {BluetoothProfile} from './interface';

/**
 * How long to scan for bluetooth peripherals, in seconds.
 */
export const SCAN_TIME = 5;
/**
 * Whether or not to allows duplicate devices
 */
export const ALLOW_DUPLICATE_DEVICES = false;
/**
 * How often to read BLE characteristic notifications
 */
export const UPDATE_INTERVAL_MS = 250;

export const _profiles = {
  femfit: FEMFIT_PROFILE,
} as const;

export type DeviceType = keyof typeof _profiles;

/**
 * Record of supported bluetooth profiles
 */
export const BLE_PROFILES = _profiles as Record<DeviceType, BluetoothProfile>;

/**
 * Record of supported peripherals
 */
export const PERIPHERAL_MAP: Record<string, DeviceType> = {
  [FEMFIT_PROFILE.batteryService.service.uuid]: 'femfit',
};
