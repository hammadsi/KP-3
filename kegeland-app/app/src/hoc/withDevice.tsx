import {capitalize} from 'lodash';
import React, {useEffect, useState, ComponentType} from 'react';

import NoDevicePopup from '~components/NoDevicePopup';
import {DeviceType} from '~constants/bluetooth';
import useDeviceSelector from '~hooks/useDeviceSelector';
import {BluetoothDevice} from '~state/ducks/bluetooth/bluetooth.interface';

export type WithDeviceContext = {
  device: BluetoothDevice | undefined;
};

/**
 * Higher-order component withDevice.
 * Wraps the component with a device selector used to select a connected
 * device from the redux state. The selected device will be passed to the sub-component
 * as a prop
 * @param deviceType the type of device to select
 * @param Component the component to wrap
 */
const withDevice =
  <P extends WithDeviceContext>(
    deviceType: DeviceType,
    Component: ComponentType<P>,
  ): ComponentType<Omit<P, keyof WithDeviceContext>> =>
  (props) => {
    const [deviceConnected, setDeviceConnected] = useState<boolean>(true);
    const device = useDeviceSelector(deviceType);
    useEffect(() => {
      setDeviceConnected(!!(device && device.state === 'connected'));
    }, [device]);

    const toggle = () => {
      setDeviceConnected(!deviceConnected);
    };

    // Set props for passing to the sub-component
    const mapStateToProps: any = {
      device,
    };

    return (
      <>
        <Component {...props} {...mapStateToProps} />
        <NoDevicePopup
          deviceName={capitalize(deviceType)}
          visible={!deviceConnected}
          onDismiss={toggle}
        />
      </>
    );
  };

export default withDevice;
