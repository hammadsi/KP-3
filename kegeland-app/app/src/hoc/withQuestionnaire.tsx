import React, {ComponentType, useEffect} from 'react';

import QuestionnaireModal from '~components/QuestionnaireModal';
import {DeviceType} from '~constants/bluetooth';
import useCurrentSession from '~hooks/useCurrentSession';
import useQuestionnaire from '~hooks/useQuestionnaire';
import {Answer} from '~state/ducks/questions/questions.interface';
import {ExerciseSession} from '~state/ducks/session/session.interface';

export type WithQuestionnaireContext = {
  session: ExerciseSession;
  answers: Answer[];
  hasQuestionnaire: boolean;
  loading: boolean;
  openQuestionnaire: () => void;
};

/**
 * Higher-order component withQuestionnaire.
 * Wraps the component with questionnaire-functionality.
 * Will fetch the questionnaire based on the device type if the signed in user
 * has a questionnaire assigned for the specific device.
 * @param deviceType the type of device to select
 * @param Component the component to wrap
 */
const withQuestionnaire =
  <P extends WithQuestionnaireContext>(
    deviceType: DeviceType,
    Component: ComponentType<P>,
  ): ComponentType<Omit<P, keyof WithQuestionnaireContext>> =>
  (props) => {
    const questionnaire = useQuestionnaire(deviceType);
    const session = useCurrentSession(questionnaire.hasQuestionnaire);

    // Set props for passing to the sub-component
    const mapStateToProps: any = {
      session: session.session,
      answers: questionnaire.answers,
      hasQuestionnaire: questionnaire.hasQuestionnaire,
      loading: questionnaire.loading || session.loading,
      openQuestionnaire: questionnaire.open,
    };

    useEffect(() => {
      // Open questionnaire if a questionnaire is assigned and 1 out of 2 answers has been answered.
      if (
        session.session &&
        questionnaire.hasQuestionnaire &&
        questionnaire.answers.length === 1
      ) {
        questionnaire.open();
      }
    }, [session.session]);

    /**
     * Upload the session if state allows for it.
     * Requires 2 answers and a session if questionnaire is enabled. Will
     * only require a session if not.
     */
    useEffect(() => {
      if (session.session && !questionnaire.visible && !session.isUploading) {
        session.upload();
      }
    }, [session.session, questionnaire.submit]);

    return (
      <>
        <Component {...props} {...mapStateToProps} />
        <QuestionnaireModal
          onSubmit={questionnaire.submit}
          visible={questionnaire.visible}
          questionnaire={questionnaire.questionnaire}
        />
      </>
    );
  };

export default withQuestionnaire;
