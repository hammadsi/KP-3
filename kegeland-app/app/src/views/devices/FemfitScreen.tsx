import React, {useEffect, useState} from 'react';
import {SafeAreaView, ScrollView, StyleSheet} from 'react-native';
import {
  ActivityIndicator,
  Button,
  Card,
  Paragraph,
  Text,
} from 'react-native-paper';

import FemfitGame from '~lib/femfit/game';
import {DeviceScreenProps} from '~routes/interface';
import {ExerciseScheme} from '~lib/femfit/game/interface';
import exercises from '~lib/femfit/game/exercises.json';
import Popup from '~components/Popup';
import {getEstimatedExerciseDuration} from '~lib/femfit/game/utils';
import PageWrapper from '~components/PageWrapper';
import {WithDeviceContext} from '~hoc/withDevice';
import {WithQuestionnaireContext} from '~hoc/withQuestionnaire';
import useAppSelector from '~hooks/useAppSelector';

export type FemfitScreenProps = DeviceScreenProps<'Femfit'> &
  WithDeviceContext &
  WithQuestionnaireContext;

/**
 * FemfitScreen. Screen for exercising with the femfit-sensor.
 * The screen will display a selection of exercises, and initiate a game once
 * an exercise is selected.
 * @see {@link FemfitScreenProps}
 */
const FemfitScreen: React.FC<FemfitScreenProps> = ({
  navigation,
  session,
  answers,
  device,
  hasQuestionnaire,
  loading,
  openQuestionnaire,
}) => {
  const [exercise, setExercise] = useState<ExerciseScheme>();
  const [modalItem, setModalItem] = useState<ExerciseScheme>();
  const {currentSession} = useAppSelector((state) => state.session);

  /**
   * Clear the state
   */
  const clear = () => {
    setExercise(undefined);
    setModalItem(undefined);
  };

  /**
   * Clear the state once the current session is undefined
   */
  useEffect(() => {
    if (!currentSession) {
      clear();
    }
  }, [currentSession]);

  /**
   * Toggle a details modal for selected exercise
   * @param item the exericse
   */
  const toggleDetails = (item?: ExerciseScheme) => {
    if (!item) {
      setModalItem(undefined);
    } else {
      setModalItem(item);
    }
  };

  /**
   * Select an exercise
   * Will trigger a questionnaire if enabled, and initiate the game
   * @param item the exercise
   */
  const selectExercise = (item: ExerciseScheme) => {
    setExercise(item);
    if (hasQuestionnaire) {
      openQuestionnaire();
    }
  };

  /**
   * Boolean function to decide whether or not the state allows for the game
   * to be rendered.
   * @returns true if game should be rendered
   */
  const shouldRenderGame = () => {
    if (device && exercise && !loading) {
      if (hasQuestionnaire) {
        return session === undefined && answers.length === 1;
      }
      return true;
    }
    return false;
  };

  const render = () => {
    if (shouldRenderGame()) {
      return (
        <SafeAreaView style={styles.wrapper}>
          <FemfitGame
            device={device!}
            exercise={exercise!}
            navigation={navigation}
          />
        </SafeAreaView>
      );
    } else if (loading) {
      return (
        <SafeAreaView style={styles.loader}>
          <ActivityIndicator animating={true} />
        </SafeAreaView>
      );
    }
    return (
      <PageWrapper
        title="Select exercise"
        contentSize="medium"
        testID="FemfitScreen">
        <ScrollView style={styles.exerciseList}>
          {(exercises as ExerciseScheme[]).map((item, idx) => (
            <Card style={styles.item} key={idx}>
              <Card.Content style={styles.itemContent}>
                <Card.Title
                  title={item.name}
                  subtitle={item.description}
                  subtitleNumberOfLines={2}
                />
                <Card.Actions style={styles.actions}>
                  <Button onPress={() => toggleDetails(item)}>Details</Button>
                  <Button onPress={() => selectExercise(item)} mode="contained">
                    Start exercise
                  </Button>
                </Card.Actions>
              </Card.Content>
            </Card>
          ))}
          <Popup
            title="Details"
            visible={modalItem !== undefined}
            contentContainerStyle={styles.detailPopup}
            onDismiss={() => toggleDetails(undefined)}
            actions={
              <Button onPress={() => toggleDetails(undefined)}>Close</Button>
            }>
            <Text style={styles.detailTitle}>{modalItem?.name}</Text>
            <ScrollView>
              <Paragraph>{modalItem?.description}</Paragraph>
            </ScrollView>
            {modalItem?.data && (
              <Text style={styles.detailDuration}>
                Estimated duration:&nbsp;
                <Text>{getEstimatedExerciseDuration(modalItem)} seconds</Text>
              </Text>
            )}
          </Popup>
        </ScrollView>
      </PageWrapper>
    );
  };

  return render();
};

const styles = StyleSheet.create({
  wrapper: {flex: 1, position: 'relative'},
  loader: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  exerciseList: {
    marginTop: 16,
  },
  item: {
    marginBottom: 16,
  },
  itemContent: {
    padding: 0,
    paddingVertical: 4,
    paddingHorizontal: 4,
  },
  detailPopup: {
    width: '90%',
  },
  detailTitle: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  detailDuration: {
    marginTop: 8,
    fontWeight: 'bold',
  },
  actions: {
    marginTop: 4,
    justifyContent: 'space-between',
  },
});

export default FemfitScreen;
