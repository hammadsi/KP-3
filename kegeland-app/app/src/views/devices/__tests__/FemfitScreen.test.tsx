import {merge} from 'lodash';
import React from 'react';

import {WithDeviceContext} from '~hoc/withDevice';
import {WithQuestionnaireContext} from '~hoc/withQuestionnaire';
import {BluetoothState} from '~state/ducks/bluetooth/bluetooth.interface';
import {initialState} from '~state/ducks/bluetooth/bluetooth.reducer';
import {deviceMock} from '~state/ducks/__mocks__/bluetooth.mocks';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';
import {render} from '~utils/test-utils';

import FemfitScreen from '../FemfitScreen';

describe('Test FemfitScreen-component', () => {
  it('should render correctly', () => {
    const hocProps: WithDeviceContext & WithQuestionnaireContext = {
      answers: [],
      device: deviceMock,
      questionnaireEnabled: true,
      questionnaireLoading: false,
      toggleQuestionnaire: jest.fn(),
    };

    const props: any = {
      navigation: {
        navigate: jest.fn(),
      },
      ...hocProps,
    };

    const authState: BluetoothState = {
      ...initialState,
      connectedDevices: {[deviceMock.id]: deviceMock},
    };
    const store = mockStore(merge({auth: authState}, initialStore));
    const component = <FemfitScreen {...props} />;
    const tree = render(component, {wrapperProps: {store}});
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
