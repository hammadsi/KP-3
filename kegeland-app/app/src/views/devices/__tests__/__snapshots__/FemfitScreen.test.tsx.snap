// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test FemfitScreen-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    edges={
      Array [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      Array [
        Object {
          "height": "100%",
          "marginTop": 8,
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 16,
        },
        undefined,
      ]
    }
    testID="FemfitScreen"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Array [
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "500",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "fontSize": 20,
                "letterSpacing": 0.15,
                "lineHeight": 30,
                "marginVertical": 2,
              },
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
            ],
          ],
        ]
      }
    >
      Select exercise
    </Text>
    <RCTScrollView
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <View>
        <View
          collapsable={false}
          nativeID="animatedComponent"
          style={
            Object {
              "backgroundColor": "#ffffff",
              "borderColor": "rgba(0, 0, 0, 0.12)",
              "borderRadius": 15,
              "elevation": 1,
              "marginBottom": 16,
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 0.5,
                "width": 0,
              },
              "shadowOpacity": 0.24,
              "shadowRadius": 0.75,
            }
          }
        >
          <View
            accessibilityState={
              Object {
                "disabled": true,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flexGrow": 1,
                "flexShrink": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "paddingHorizontal": 16,
                  },
                  Object {
                    "paddingVertical": 16,
                  },
                  Object {
                    "padding": 0,
                    "paddingHorizontal": 4,
                    "paddingVertical": 4,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "paddingLeft": 16,
                    },
                    Object {
                      "minHeight": 72,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "flexDirection": "column",
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "fontFamily": "System",
                            "fontWeight": "500",
                            "writingDirection": "ltr",
                          },
                          Array [
                            Object {
                              "fontSize": 20,
                              "letterSpacing": 0.15,
                              "lineHeight": 30,
                              "marginVertical": 2,
                            },
                            Array [
                              Object {
                                "minHeight": 30,
                                "paddingRight": 16,
                              },
                              Object {
                                "marginBottom": 0,
                              },
                              undefined,
                            ],
                          ],
                        ],
                      ]
                    }
                  >
                    Exercise 1
                  </Text>
                  <Text
                    numberOfLines={2}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.54)",
                            "fontFamily": "System",
                            "fontWeight": "400",
                            "writingDirection": "ltr",
                          },
                          Array [
                            Object {
                              "fontSize": 12,
                              "letterSpacing": 0.4,
                              "lineHeight": 20,
                              "marginVertical": 2,
                            },
                            Array [
                              Object {
                                "marginVertical": 0,
                                "minHeight": 20,
                                "paddingRight": 16,
                              },
                              undefined,
                            ],
                          ],
                        ],
                      ]
                    }
                  >
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin dictum vitae nibh id posuere. Integer varius ac dolor non bibendum. Pellentesque vestibulum aliquam tincidunt. Suspendisse congue lobortis risus, ac luctus sem suscipit sed. Ut et tincidunt lacus. Suspendisse placerat ornare velit vel lobortis. Cras sagittis arcu pulvinar, venenatis ex in, dictum sapien. Curabitur tempus felis eget arcu facilisis interdum
                  </Text>
                </View>
                <View />
              </View>
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                      "padding": 8,
                    },
                    Object {
                      "justifyContent": "space-between",
                      "marginTop": 4,
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderColor": "transparent",
                      "borderRadius": 15,
                      "borderStyle": "solid",
                      "borderWidth": 0,
                      "elevation": 0,
                      "minWidth": "auto",
                      "shadowColor": "#000000",
                      "shadowOffset": Object {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                    }
                  }
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      Object {
                        "disabled": false,
                      }
                    }
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Object {
                          "borderRadius": 15,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          undefined,
                        ]
                      }
                    >
                      <Text
                        numberOfLines={1}
                        selectable={false}
                        style={
                          Array [
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontWeight": "400",
                            },
                            Object {
                              "textAlign": "left",
                            },
                            Array [
                              Object {
                                "letterSpacing": 1,
                                "marginHorizontal": 16,
                                "marginVertical": 9,
                                "textAlign": "center",
                              },
                              Object {
                                "marginHorizontal": 8,
                              },
                              Object {
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "#00b0ff",
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              Object {
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                        Details
                      </Text>
                    </View>
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "backgroundColor": "#00b0ff",
                      "borderColor": "transparent",
                      "borderRadius": 15,
                      "borderStyle": "solid",
                      "borderWidth": 0,
                      "elevation": 2,
                      "minWidth": "auto",
                      "shadowColor": "#000000",
                      "shadowOffset": Object {
                        "height": 0.75,
                        "width": 0,
                      },
                      "shadowOpacity": 0.24,
                      "shadowRadius": 1.5,
                    }
                  }
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      Object {
                        "disabled": false,
                      }
                    }
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Object {
                          "borderRadius": 15,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          undefined,
                        ]
                      }
                    >
                      <Text
                        numberOfLines={1}
                        selectable={false}
                        style={
                          Array [
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontWeight": "400",
                            },
                            Object {
                              "textAlign": "left",
                            },
                            Array [
                              Object {
                                "letterSpacing": 1,
                                "marginHorizontal": 16,
                                "marginVertical": 9,
                                "textAlign": "center",
                              },
                              Object {
                                "marginHorizontal": 8,
                              },
                              Object {
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "#000000",
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              Object {
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                        Start exercise
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          nativeID="animatedComponent"
          style={
            Object {
              "backgroundColor": "#ffffff",
              "borderColor": "rgba(0, 0, 0, 0.12)",
              "borderRadius": 15,
              "elevation": 1,
              "marginBottom": 16,
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 0.5,
                "width": 0,
              },
              "shadowOpacity": 0.24,
              "shadowRadius": 0.75,
            }
          }
        >
          <View
            accessibilityState={
              Object {
                "disabled": true,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flexGrow": 1,
                "flexShrink": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "paddingHorizontal": 16,
                  },
                  Object {
                    "paddingVertical": 16,
                  },
                  Object {
                    "padding": 0,
                    "paddingHorizontal": 4,
                    "paddingVertical": 4,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "paddingLeft": 16,
                    },
                    Object {
                      "minHeight": 72,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "flexDirection": "column",
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "fontFamily": "System",
                            "fontWeight": "500",
                            "writingDirection": "ltr",
                          },
                          Array [
                            Object {
                              "fontSize": 20,
                              "letterSpacing": 0.15,
                              "lineHeight": 30,
                              "marginVertical": 2,
                            },
                            Array [
                              Object {
                                "minHeight": 30,
                                "paddingRight": 16,
                              },
                              Object {
                                "marginBottom": 0,
                              },
                              undefined,
                            ],
                          ],
                        ],
                      ]
                    }
                  >
                    Exercise 2
                  </Text>
                  <Text
                    numberOfLines={2}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.54)",
                            "fontFamily": "System",
                            "fontWeight": "400",
                            "writingDirection": "ltr",
                          },
                          Array [
                            Object {
                              "fontSize": 12,
                              "letterSpacing": 0.4,
                              "lineHeight": 20,
                              "marginVertical": 2,
                            },
                            Array [
                              Object {
                                "marginVertical": 0,
                                "minHeight": 20,
                                "paddingRight": 16,
                              },
                              undefined,
                            ],
                          ],
                        ],
                      ]
                    }
                  >
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin dictum vitae nibh id posuere. Integer varius ac dolor non bibendum. Pellentesque vestibulum aliquam tincidunt. Suspendisse congue lobortis risus, ac luctus sem suscipit sed. Ut et tincidunt lacus. Suspendisse placerat ornare velit vel lobortis. Cras sagittis arcu pulvinar, venenatis ex in, dictum sapien. Curabitur tempus felis eget arcu facilisis interdum
                  </Text>
                </View>
                <View />
              </View>
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                      "padding": 8,
                    },
                    Object {
                      "justifyContent": "space-between",
                      "marginTop": 4,
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderColor": "transparent",
                      "borderRadius": 15,
                      "borderStyle": "solid",
                      "borderWidth": 0,
                      "elevation": 0,
                      "minWidth": "auto",
                      "shadowColor": "#000000",
                      "shadowOffset": Object {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                    }
                  }
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      Object {
                        "disabled": false,
                      }
                    }
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Object {
                          "borderRadius": 15,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          undefined,
                        ]
                      }
                    >
                      <Text
                        numberOfLines={1}
                        selectable={false}
                        style={
                          Array [
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontWeight": "400",
                            },
                            Object {
                              "textAlign": "left",
                            },
                            Array [
                              Object {
                                "letterSpacing": 1,
                                "marginHorizontal": 16,
                                "marginVertical": 9,
                                "textAlign": "center",
                              },
                              Object {
                                "marginHorizontal": 8,
                              },
                              Object {
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "#00b0ff",
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              Object {
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                        Details
                      </Text>
                    </View>
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "backgroundColor": "#00b0ff",
                      "borderColor": "transparent",
                      "borderRadius": 15,
                      "borderStyle": "solid",
                      "borderWidth": 0,
                      "elevation": 2,
                      "minWidth": "auto",
                      "shadowColor": "#000000",
                      "shadowOffset": Object {
                        "height": 0.75,
                        "width": 0,
                      },
                      "shadowOpacity": 0.24,
                      "shadowRadius": 1.5,
                    }
                  }
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      Object {
                        "disabled": false,
                      }
                    }
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Object {
                          "borderRadius": 15,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          undefined,
                        ]
                      }
                    >
                      <Text
                        numberOfLines={1}
                        selectable={false}
                        style={
                          Array [
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontWeight": "400",
                            },
                            Object {
                              "textAlign": "left",
                            },
                            Array [
                              Object {
                                "letterSpacing": 1,
                                "marginHorizontal": 16,
                                "marginVertical": 9,
                                "textAlign": "center",
                              },
                              Object {
                                "marginHorizontal": 8,
                              },
                              Object {
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "#000000",
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              Object {
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                        Start exercise
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RNCSafeAreaView>
</View>
`;
